### ISSUE TITLE
`trivial_casts` warning misfires when casting from reference -> pointer -> void pointer

### ISSUE THREAD

**Issue by stouset** [rust-lang/rust#52200]
When casting from a reference to a pointer to a void pointer (necessary sometimes when doing C FFI), the `trivial_casts` warning fires, despite this being an necessary cast.

```rust
#![warn(trivial_casts)]

extern crate libc;

fn main() {
    let x : libc::c_int  = 4;
    let y : *libc::c_void = &mut x
        as *libc::c_int
        as *libc::c_void;
}
```

You can work around this by using coercion to assign the intermediate type to a variable, but this shouldn't be necessary.

**Comment by scottmcm** [comment]
Given that the warning explicitly says

> Cast can be replaced by coercion, this might require type ascription or a temporary variable

It seems like this is by design?

**Comment by oli-obk** [comment]
The example code is not valid rust code.

Maybe the lint triggers on casting away mutability?

Can you minimize your example to have no dependencies?

**Comment by stouset** [comment]
Whoops, sorry. My original code used mutable references and I attempted to remove that part to make an even more minimal example. I should have run it back through `rustc` before posting. I've edited the original example.

**Comment by stouset** [comment]
@scottmcm I did note that you can work around this by assigning to an intermediate variable, but this warning is supposed to fire on trivial (e.g., unnecessary) casts. The intermediate cast is necessary and unavoidable, since to step from a `&Foo` to a `*const c_void` you must first pass through `*const Foo`.

**Comment by scottmcm** [comment]
> this warning is supposed to fire on trivial (e.g., unnecessary) casts

It's doing exactly that, though, so I'm confused what the problem is.  Another cast elsewhere in the program doesn't make the first cast non-trivial.

> However, such a cast can't be elided (without assigning to a temporary variable as a workaround) when it's just a stepping stone to a void pointer for use in FFI.

Right, but what's wrong with a temporary variable, given that the warning explicitly suggests one?

You can always do something like this, if you like more-inline:
```rust
    let x : libc::c_int = 4;
    let _y : *const libc::c_void = { let t: *const libc::c_int = &x; t}
        as *const libc::c_void;
```

Or even nicer, hide the coercion in a function call like
```rust
fn to_c_void_ptr<T>(p: *const T) -> *const libc::c_void {
    p as _
}

fn main() {
    let x : libc::c_int = 4;
    let y : *const libc::c_void = to_c_void_ptr(&x);
}
```

**Comment by stouset** [comment]
> Right, but what's wrong with a temporary variable, given that the warning explicitly suggests one?

Just like your other suggestions, It's a workaround, and one that shouldn't be necessary given that the cast is strictly necessary for the given task trying to be accomplished. The point of warnings in the first place is to alert a developer to problems (whether logical or stylistic) in their code. This warning is going off in a situation where there *is no problem*, and the original solution is in fact the simplest, easiest, and most straightforward one.

Warnings should not steer developers toward writing more complicated code than is actually necessary. This warning, for example, normally fires when explicit casts are provided but they can simply be removed, resulting in simpler code. In this case, the warning fires in a way that recommends unnecessarily *adding* complication to code: all of your workarounds are clearly more complicated—and unnecessarily so—than a simple double-cast.

Your argument is fundamentally based on circular reasoning: the provided code is triggering the warning because the warning currently triggers on that code. My argument is that it doesn't *need* to trigger in this case (the cast is not extraneous), and that doing so is actively (if very slightly) harmful, since it advises developers to make their code more complicated than is actually necessary.

**Comment by CJKay** [comment]
Any update on this? It triggers on every cast where you need the address as `usize`:

```
warning: trivial cast: `&mut T` as `*mut T`
   --> src\lib.rs:200:22
    |
200 |         OpParam::new(op as *mut T as usize)
    |                      ^^^^^^^^^^^^
    |
note: lint level defined here
   --> src\lib.rs:101:9
    |
101 | #![warn(trivial_casts)]
    |         ^^^^^^^^^^^^^
    = help: cast can be replaced by coercion; this might require a temporary variable
```

**Comment by kornelski** [comment]
`as *mut T as usize` is still misfiring.
